broker/aws/impl.go

```diff
@@ -0,0 +1,254 @@
+	accountID, err := b.accountIDFromName(accountName)
+	if err != nil {
+		return nil, err
+	}
+	roleArn := fmt.Sprintf("arn:aws:iam::%s:role/%s", accountID, roleName)
+	assumeRoleInput := sts.AssumeRoleInput{
+		DurationSeconds: &durationSeconds,
+		RoleArn:         &roleArn,
+		RoleSessionName: &roleSessionName,
+	}
+	return stsClient.AssumeRole(&assumeRoleInput)
+}
+
+func (b *Broker) withSessionGetAWSRoleList(validSession *session.Session) ([]string, error) {
+	iamClient := iam.New(validSession)
+	var maxItems int64
+	maxItems = 500
+	listRolesInput := iam.ListRolesInput{MaxItems: &maxItems}
+	listRolesOutput, err := iamClient.ListRoles(&listRolesInput)
+	if err != nil {
+		b.logger.Printf("list profile failed")
+		return nil, err
+	}
+	//TODO, check for truncation!
+	b.logger.Debugf(2, "listrolesoutput =%v", listRolesOutput)
+
+	var roleNames []string
+	for _, role := range listRolesOutput.Roles {
+		roleNames = append(roleNames, *role.RoleName)
+	}
+	sort.Strings(roleNames)
+	b.logger.Debugf(1, "roleNames(v =%v)", roleNames)
+	return roleNames, nil
+}
+
+func (b *Broker) masterGetAWSRolesForAccount(accountName string) ([]string, error) {
+	assumeRoleOutput, err := b.withProfileAssumeRole(accountName, masterAWSProfileName, masterRoleName)
+	if err != nil {
+		b.logger.Printf("cannot assume role for account %s, err=%s", accountName, err)
+		return nil, err
+	}
+	b.logger.Printf("assume role success for account=%s, roleoutput=%v", accountName, assumeRoleOutput)
+
+	//TODO check region from account config
+	region := "us-east-1"
+	sessionCredentials := credentials.NewStaticCredentials(*assumeRoleOutput.Credentials.AccessKeyId,
+		*assumeRoleOutput.Credentials.SecretAccessKey, *assumeRoleOutput.Credentials.SessionToken)
+	assumedSession := session.Must(session.NewSession(aws.NewConfig().WithCredentials(sessionCredentials).WithRegion(region)))
+
+	return b.withSessionGetAWSRoleList(assumedSession)
+}
+
+func (b *Broker) getAWSRolesForAccount(accountName string) ([]string, error) {
+	b.logger.Debugf(1, "top of getAWSRolesForAccount for account =%s", accountName)
+	accountRoles, err := b.masterGetAWSRolesForAccount(accountName)
+	if err == nil {
+		return accountRoles, nil
+	}
+	b.logger.Printf("Doing fallback for accountName=%s", accountName)
+	// Master role does not work, try fallback with direct account
+	profileName := accountName
+	sessionCredentials, region, err := b.getCredentialsFromProfile(profileName)
+	if err != nil {
+		return nil, err
+	}
+
+	if sessionCredentials == nil {
+		return nil, errors.New(fmt.Sprintf("no valid profile=%s", profileName))
+	}
+
+	// This is strange, no error calling?
+	accountSession := session.Must(session.NewSession(aws.NewConfig().WithCredentials(sessionCredentials).WithRegion(region)))
+	return b.withSessionGetAWSRoleList(accountSession)
+}
+
+func stringIntersectionNoDups(set1, set2 []string) (intersection []string) {
+	for _, v1 := range set1 {
+		for _, v2 := range set2 {
+			if v1 == v2 {
```

Comment:
This is an O(M*N) algorithm :-(
Please use a map

Reply:
M and N are both small in these cases (5-10), so it does not really matter. If I make it into hashes I get-> O(n) + O(m) with much larger constants. (also this will start to get messy on this function due to casing on later patches). Lets mark it as a todo?

Reply:
OK. TODO for later.